src/exposure.jl

```diff
@@ -0,0 +1,177 @@
+At every time step the pesticide is degradated at rate `degradationrate`
+
+# Keyword Arguments
+
+- `crop`: Crop field. May be a `Number`, an [`Aux`](@ref) array or another [`Grid`](@ref).
+Treatment is applied when `crop > zero(crop) || return zero(pesticide) `.
+- `rotationsize`: Total number of type in rotation.  May be a `Number`, an [`Aux`](@ref) array or
+another [`Grid`](@ref).
+- `rotationindex`: Index of the treatment applied. A type compatible with `rotationsize`.
+- `pulselevel`: Pulse level of treatment. May be a `Number`, an [`Aux`](@ref) array or
+another [`Grid`](@ref).
+- `popthreshold`: Threshold population level triggering the treatment release. May be a
+ `Number`, an [`Aux`](@ref) array or another [`Grid`](@ref) of the same type as the
+ `population` layer (Read is a Tuple(pesticide, population)) since used in the comparison:
+ `population < rule.popthreshold`.
+- `degradationrate`: Rate of degradation of the treatment. May be a `Number`, an
+ [`Aux`](@ref) array or another [`Grid`](@ref) of a type compatible with `pulselevel`.
+- `timestep`: Time step for the exposure, in a type compatible with the simulation `tspan`.
+
+Pass grid `Symbol`s to `R` or both `R` and `W` type parameters to use to specific grids.
+Read: Tuple(pesticide, population, rotation)
+Written: Tuple(pesticide, rotation)
+"""
+struct RotationExposure{R,W,CP,RS,RN, PL,PT,DR,TS,S} <: Exposure{R,W}
+    crop::CP
+    rotationsize::RS
+    rotationindex::RN
+    pulselevel::PL
+    popthreshold::PT
+    degradationrate::DR
+    timestep::TS
+    nsteps::S
+end
+RotationExposure{R,W}(;
+    crop=CROP_OCCUPANCY,
+    rotationsize=ROTATION_SIZE,
+    rotationindex=ROTATION_NUMBER,
+    pulselevel=PULSE_LEVEL,
+    popthreshold=POPULATION_THRESHOLD,
+    degradationrate=DEGRADATION_RATE,
+    timestep,
+) where {R,W} = RotationExposure{R,W}(crop, rotationsize, rotationindex, pulselevel, popthreshold, degradationrate, timestep, nothing)
+
+precalcrule(rule::RotationExposure, data) = precalc_timestep(rule, data)
+
+@inline function applyrule(data, rule::RotationExposure, (pesticide, population, rotation), I)
+    crop = get(data, rule.crop, I...) 
+    crop > zero(crop) || return (zero(pesticide), rotation)
+    pulse = get(data, rule.pulselevel, I...)
+    z = get(data, rule.popthreshold, I...)
+    rt = get(data, rule.degradationrate, I...) * rule.nsteps
+    if  ( population >= z &&
+        mod(_rotationstep(rotation), rule.rotationsize) == mod(rule.rotationindex, rule.rotationsize) &&
+        _timestep(rotation) != currentframe(data) )
+        # println("IN $I")
+        @fastmath (
+            pesticide * exp(- rt) + pulse,
+            _updatestep(rotation + 1, currentframe(data))
+        )
+    else
+        # println("OUT $I")
+        @fastmath (
+            pesticide * exp(- rt),
+            rotation
+        )
+    end
+end
+
+struct RotationStruct{RS,TS}
+    rotationStep::RS
```

Comment:
yes, this `timeStep` is different to `timestep`used elsewhere. It is used to keep the date at which the treatment has been applied, I rename as `datestep`.

Reply:
maybe not `datestep`, as a `step` is normally used for intervals between things. Is it more like `timestamp` ? You also don't know it will be a date explicitly, but you always know it's at least the time dimension.